common/http/platform/beast/http_resumer.cpp

```diff
@@ -0,0 +1,454 @@
+	user_header_handler_(exp_resp);
+
+	io::WriterPtr user_writer = resp->GetBodyWriter();
+	if (!user_writer) {
+		auto err = MakeError(
+			DownloadResumerError, "The user did not set a BodyWriter to write the data into");
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		return;
+	}
+	client_.state_.user_body_writer = user_writer;
+}
+void HeaderHandlerFunctor::HandleNextResponse(ExpectedIncomingResponsePtr exp_resp) {
+	// Subsequent responses shall not call the user handler. If an error occurs, either
+	// schedule the next AsyncCall directly or save it and let the body handler forward
+	// it to the user
+	if (!exp_resp) {
+		client_.logger_.Warning(exp_resp.error().String());
+
+		auto err = ScheduleNextResumeRequest();
+		if (err != error::NoError) {
+			client_.logger_.Error(err.String());
+			client_.Cancel();
+			user_body_handler_(expected::unexpected(err));
+		}
+		return;
+	}
+	auto resp = exp_resp.value();
+
+	auto exp_content_range_header = resp->GetHeader("Content-Range");
+	if (!exp_content_range_header) {
+		client_.logger_.Error(exp_content_range_header.error().String());
+		client_.Cancel();
+		client_.state_.err = exp_content_range_header.error();
+		return;
+	}
+
+	auto exp_content_range = ParseRangeHeader(exp_content_range_header.value());
+	if (!exp_content_range) {
+		client_.logger_.Error(exp_content_range.error().String());
+		client_.Cancel();
+		client_.state_.err = exp_content_range.error();
+		return;
+	}
+
+	auto content_range = exp_content_range.value();
+	if (content_range.size != 0 && content_range.size != client_.state_.content_length) {
+		auto err = MakeError(
+			DownloadResumerError,
+			"Size of artifact changed after download was resumed (expected "
+				+ to_string(client_.state_.content_length) + ", got "
+				+ to_string(content_range.size) + ")");
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		client_.state_.err = err;
+		return;
+	}
+	if ((content_range.range_end != client_.state_.content_length - 1)
+		|| (content_range.range_start > client_.state_.offset)) {
+		auto err = MakeError(
+			DownloadResumerError,
+			"HTTP server did not return expected range. Expected "
+				+ to_string(client_.state_.offset) + "-"
+				+ to_string(client_.state_.content_length - 1) + ", got "
+				+ to_string(content_range.range_start) + "-" + to_string(content_range.range_end));
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		client_.state_.err = err;
+	} else if (content_range.range_start < client_.state_.offset) {
+		// The server resumed from an earlier offset, but we can still use the data
+		auto buffer = client_.state_.buffer.get();
+		auto overriden = client_.state_.offset - content_range.range_start;
+		client_.logger_.Warning(
+			"HTTP server returned a range with an earlier start. Expected "
+			+ to_string(client_.state_.offset) + ", got " + to_string(content_range.range_start)
+			+ ". The last " + to_string(overriden) + " bytes will be overriden");
```

Comment:
You mean "The first x bytes", not the last, or? Also, I would maybe say "discarded", rather than overridden.

Reply:
I mean the last x bytes of what already has been received in previous HTTP responses. Overridden is better in the sense that the new content is overriding the last x bytes that were there.

This is a bit different than the [golang implementation](https://github.com/mendersoftware/mender/blob/d9010526d35d3ac861ea1e4210d36c2fef748ef8/client/update_resumer.go#L157) because here I don't have control of the incoming stream to "consume" the first x bytes (it is controlled by the underlying client) so instead I remove the last x bytes of my buffer and concatenate the buffer from there.

Once clarified, do you have a suggestion on how to better rewrite the warning message?

Reply:
Yeah, I see what you mean now. You are always operating within one buffer block, so you can actually overwrite bytes you already received. Then the message is good. 👍